thirdparty/patches/opencensus-cpp-shutdown-api.patch

```diff
@@ -20,44 +21,60 @@ diff --git opencensus/stats/internal/stats_exporter.cc opencensus/stats/internal
 +    t_.join();
 +  }
  }
-
+ 
  void StatsExporterImpl::RunWorkerLoop() {
    absl::Time next_export_time = GetNextExportTime();
 -  while (true) {
-+  while (thread_started_) {
++  bool thread_started = false;
++  {
++    absl::MutexLock l(&mu_);
++    bool thread_started = thread_started_;
++  }
++  while (thread_started) {
      // SleepFor() returns immediately when given a negative duration.
      absl::SleepFor(next_export_time - absl::Now());
      // In case the last export took longer than the export interval, we
-@@ -142,5 +153,10 @@ void StatsExporter::ClearHandlersForTesting() {
-   StatsExporterImpl::Get()->ClearHandlersForTesting();
+     // calculate the next time from now.
+     next_export_time = GetNextExportTime();
+     Export();
++    {
++      absl::MutexLock l(&mu_);
```

Comment:
It's weird. Can we mark thread_started_ in volatile?

Reply:
Can you elaborate a little more? I wonder 1. Why do you want it to be volatile? 2. Why is it weird? (I know it is weird. I don't know how to make this clean in cpp, so please let me know if you have any good suggestion). 

Btw, the reasoning for this is because `thread_start_` requires to acquire `mu_`. But we should hold a lock only when this value is changed (otherwise, there could be deadlock when `Export` is called).

Reply:
> Can you elaborate a little more? I wonder 1. Why do you want it to be volatile? 2. Why is it weird? (I know it is messy. I don't know how to make this clean in cpp, so please let me know if you have any good suggestion).
> 
> Btw, the reasoning for this is because `thread_start_` requires to acquire `mu_`. But we should hold a lock only when this value is changed (otherwise, there could be deadlock when `Export` is called).

I mean volatile property can be accessed directly without any cahched in register of CPU, so all of threads hold the true value not cached value, which doesn't make exporting anymore. Atomic flag can do this as well.

If exporte is invoked, the thread will wait unitl this call finished. So only one thing we should make sure that export would be terminated or finished eventually. Besides, no async function will be called except io service used in rpc client. Once io service of rpc client stopped, I think the exporter thread should be joined and release all. Any corner case I missed?

Reply:
> > Can you elaborate a little more? I wonder 1. Why do you want it to be volatile? 2. Why is it weird? (I know it is messy. I don't know how to make this clean in cpp, so please let me know if you have any good suggestion).
> > Btw, the reasoning for this is because `thread_start_` requires to acquire `mu_`. But we should hold a lock only when this value is changed (otherwise, there could be deadlock when `Export` is called).
> 
> I mean volatile property can be accessed directly without any cahched in register of CPU, so all of threads hold the true value not cached value, which makes no exporting anymore if thread_started changed in other thread. Atomic flag can do this as well.
> 
> If exporte is invoked, the thread will wait unitl this call finished. So only one thing we should make sure that export would be terminated or finished eventually. Besides, no async function will be called except io service used in rpc client. Once io service of rpc client stopped, I think the exporter thread should be joined and release all. Any corner case I missed?

Reply:
Ah, I see. I didn't know about volatile (only knew atomic). Good to know. I will use atomic bool in this case (since opencensus seems to already have absl dependencies)? 

> If exporte is invoked, the thread will wait unitl this call finished. So only one thing we should make sure that export would be terminated or finished eventually. Besides, no async function will be called except io service used in rpc client. Once io service of rpc client stopped, I think the exporter thread should be joined and release all. Any corner case I missed?

Nothing missing here, but it is a little bit inaccurate. I think what's happening is 
1. io_service stopped.
2. statsExporterImpl::Get()::shutdown called => This will stop the loop.
3. If export is still going on, they are all ignore because the io service stopped.
4. Shutdown waits until the last loop is done and threads are joined.

Reply:
Actually, it doesn't work after I started using atomic bool. Let me actually just merge this now, and try to improve it next week while I am working on closing deltaproducers.